File: src/delete/strategy.js
export function planPrune(schema, record) {
  // TODO schema strategy

  // TODO: handle case when value is array
  // { _: a, a: [ { _: a, a: a1 }, { _: a, a: a1 } ] }
  const { _: base, [base]: baseValue } = record;

  const trunks = schema[base].trunk;

  const trunkTablets = trunks.map((trunk) => {
    return {
      filename: `${trunk}-${base}.csv`,
      trait: baseValue,
      traitIsFirst: false,
    }
  });

  const { leaves } = schema[base];

  const leafTablets = leaves.map((leaf) => ({
    filename: `${base}-${leaf}.csv`,
    trait: baseValue,
    traitIsFirst: true,
  }));

  return [...trunkTablets, ...leafTablets];
}

File: src/dataset/delete/strategy.rs
use crate::{line::Line, Branch, Dataset, Entry, Error, Leaves, Result, Schema, Trunks};
use serde::{Deserialize, Serialize};
use std::fs;
use std::fs::File;
use std::io::prelude::*;
use std::path::{Path, PathBuf};
use temp_dir::TempDir;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Tablet {
    pub filename: String,
    pub trait_: String,
    pub trait_is_first: bool,
}

pub fn plan_delete(schema: &Schema, query: &Entry) -> Result<Vec<Tablet>> {
    let (trunks, leaves) = match schema.0.get(&query.base) {
        None => (vec![], vec![]),
        Some(Branch {
            trunks: Trunks(ts),
            leaves: Leaves(ls),
        }) => (ts.to_vec(), ls.to_vec()),
    };

    let base_value = match &query.base_value {
        None => return Err(Error::from_message("unexpected missing option")),
        Some(v) => v,
    };

    let trunk_tablets: Vec<Tablet> = trunks
        .iter()
        .map(|trunk| Tablet {
            filename: format!("{}-{}.csv", trunk, query.base),
            trait_: base_value.to_owned(),
            trait_is_first: false,
        })
        .collect();

    let leaf_tablets = leaves
        .iter()
        .map(|leaf| Tablet {
            filename: format!("{}-{}.csv", query.base, leaf),
            trait_: base_value.to_owned(),
            trait_is_first: true,
        })
        .collect();

    Ok([trunk_tablets, leaf_tablets].concat())
}
